### ISSUE TITLE
TargetFramework cannot be set in Directory.Build.props

### ISSUE THREAD

**Issue by Liversage** [stryker-mutator/stryker-net#3144]
**Describe the bug**

When `<TargetFramework>` is set in `Directory.Build.props` Stryker doesn't work.

**Logs**

I have a a solution folder with a `Directory.Build.props` that specifies `<TargetFramework>net9.0</TargetFramework>` and there are no `<TargetFramework>` in the `.csproj` files. The project to mutate and the unit test project are in subdirectories and the unit test project has a reference to the project to mutate. When executing `dotnet stryker --verbosity debug` in the unit test project I get the following console output:

<pre>
[13:07:16 INF] Logging enabled at level Debug
Version: 4.4.1
[13:07:16 DBG] Stryker starting, version: 4.4.1

[13:07:16 DBG] Stryker started with options: [REMOVED FOR BREVITY]
[13:07:16 INF] Analysis starting.
[13:07:16 DBG] Using C:\Users\[REDACTED]\[REDACTED].UnitTests\[REDACTED].UnitTests.csproj as test project
[13:07:16 INF] Analyzing 1 test project(s).
[13:07:16 DBG] Analyzing 0 projects.
[13:07:16 DBG] Analyzing [REDACTED].UnitTests.csproj
[13:07:22 DBG] Analysis of project [REDACTED].UnitTests.csproj succeeded.
[13:07:23 DBG] Analyzing ..\[REDACTED]\[REDACTED].csproj
[13:07:29 DBG] Analysis of project ..\[REDACTED]\[REDACTED].csproj succeeded.
[13:07:29 INF] Analysis complete.
[13:07:29 INF] Time Elapsed 00:00:13.2623456
Stryker.NET failed to mutate your project. For more information see the logs below:

No project references found. Please add a project reference to your test project and retry.
</pre>

**Expected behavior**

According to the logs Stryker correctly detects the project to mutate and should go ahead and perform the tests.

Adding `<TargetFramework>net9.0</TargetFramework>` to both `.csproj` files fixes the problem.

**Desktop**
 - OS: Windows
 - Type of project: "core"
 - Framework Version: .NET 9.0
 - Stryker Version: 4.4.1

**Additional context**

I haven't tried to locate the actual problem in the source code. Even if it's not worth fixing it might be helpful to explain this in the documentation. This is my first time using Stryker and it took me quite some time to understand why my project didn't work with Stryker despite following the clear and useful instructions in "Getting started".

**Comment by Liversage** [comment]
@dupdob Yes, that's exactly the issue. The `<TargetFramework>` is not being picked up correctly so probably not something that can be fixed by this project.

**Comment by Liversage** [comment]
> Not sure what's going on here because we use this ourselves in Stryker: https://github.com/stryker-mutator/stryker-net/blob/master/src/Directory.Build.props#L4

Perhaps `<TargetFramework>` is also set directly in the `.csproj` file or more likely the "fallback" `<TargetFramework>` used by Buildalyzer (I believe it's full framework) doesn't cause any compile errors. In my case I was using top-level statements and that didn't compile when using Stryker.

**Comment by dupdob** [comment]
Note that current Stryker releases do not support top-level statements....
This will be fixed in next release.

**Comment by willseward** [comment]
There might be something else wrong, here. We use the `--target-framework` option in a config file to override the property in the .props file. It used to work, but it seems in the later versions of the tool it doesn't. I went back and did some testing, and it seems that this behavior changed in version 4.3.0: 4.2.0 works fine, but 4.3.0 is broken.

I must say...this commit seems suspicious: https://github.com/stryker-mutator/stryker-net/commit/85a4384df9a24aa15fad55a2901988f1c042588b

**Comment by dupdob** [comment]
yes there have been a change in how `--target-framework` to increase compatibility with various project configurations (you can see PR #3039 to have details about this.

Can you explain what you mean exactly by 'used to work' and 'now it don't'? Stryker is no longer able to build the project or it uses a different framework version in latest versions?

**Comment by willseward** [comment]
Sure. We have a project that sets `<TargetFramework>`  in `Directory.Build.props`, per this issue. Version 4.2.0 would fail to build the project when simply running `dotnet stryker`, but we were able to work around that by setting the `--target-framework` option in either the cli or the config json. In version 4.3.0, that behavior changed, we the `--target-framework` no longer sets the build framework, and the build fails with this message:

> No project references found. Please add a project reference to your test project and retry.

I'm not sure what framework stryker is using to attempt the build, but the output is identical to that referenced in the issue description.

**Comment by dupdob** [comment]
Thanks for the feedback. 
> We have a project that sets <TargetFramework> in Directory.Build.props

Does this mean there is no `TargetFramework` defined in your project(s)?

I guess this is the case, because the above mentioned PR focused on solving issues when working with complex multi targeting solutions. For those, forcing the framework often resulted in failed analysis, so we switched to an improved analysis strategy while never forcing the framework.

But, it is fair to assume that if no target is present in (some?) projects, not forcing will lead to failed analysis as well.

Next steps:
1. If you confirm there is no target framework defined in your project, I will try to reproduce this and find a solution. I may ask you to provide further details to help me. But if you can share this project or any project replicating the issue it would be a great help
2. If this is no the case, having access to your project (or a project replicating) the issue would be great, if now, could you please share with me the csproj files (can be done privately if you prefer)?

**Comment by willseward** [comment]
We do define a `TargetFramework`. We do it in this manner:

*Directory.Build.props*
```xml
<Project>
  <PropertyGroup>
    <ProjectTarget>net8.0</ProjectTarget>

<!-- etc. -->
```

*.csproj*
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>$(ProjectTarget)</TargetFramework>

<!-- etc. -->
```

I cannot provide any access to the proj files, unfortunately, but I hope the above provides enough information.

**Comment by dupdob** [comment]
thanks, I think I have enough information to reproduce this. I will get back to you if I need anything more.

**Comment by dupdob** [comment]
hi
are you sure Stryker does not work with `Directory.Build.Props` ?

I did reproduce the issue when injecting this in a test project:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>$(ProjectTarget)</TargetFramework>
```
and never defining this property (`ProjectTarget`). So far so good. But the initial build kept on failing due to this property not being defined.
Long story short, I added this property to our `Directory.Build.Props` file and everything worked as expected, without needing to specify the target framework.

My assumption for this situation is now:
1) you can use latest Stryker version on your project without specifying target framework. The actual framework is the one defined in the `props` file.
2) for some reason you prefer to force this value to something else and this no longer works with the latest version.

I am asking because I realized that I did not truly reproduce your issue as I have to skip declaring the property to reproduce it, so my 'fix' has probably no value.

btw, my fix was simply to force the target framework to the configured value as a last attempt when previous analysis failed.

**Comment by rouke-broersma** [comment]
To add on the what @dupdob mentioned I am also curious why you use `<TargetFramework>$(ProjectTarget)</TargetFramework>` in the first place. If you leave out `<TargetFramework>` and set `<TargetFramework>` in your Directory.Build.Props, msbuild will automatically take this value as your target framework. No need to use msbuild variable substitution. We do this here: https://github.com/stryker-mutator/stryker-net/blob/11758103fca00e1106f5a7c23240fd5cd255f7af/src/Directory.Build.props#L4

As you can see here our csproj has _no_ target framework property: https://github.com/stryker-mutator/stryker-net/blob/master/src/Stryker.Core/Stryker.Core/Stryker.Core.csproj

And this works.

**Comment by willseward** [comment]
@rouke-broersma  Not all of our project are able to target the same framework. Analyzer projects, for example, must target netstandard2.0 for VS compatibility. 

Nonetheless, I will attempt to get a minimal repro of this issue.

**Comment by willseward** [comment]
@dupdob  I have a minimal repo that I can share.  [Repro.zip](https://github.com/user-attachments/files/20783790/Repro.zip)

Out of the box, the following fails:
```bash
> dotnet stryker -tp .\Tests\Tests.csproj -p Library\Library.csproj

   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 4.6.0

[15:30:21 INF] Analysis starting.
[15:30:21 INF] Identifying projects to mutate in Repro.sln. This can take a while.
[15:30:27 WRN] Analysis of project Tests\Tests.csproj failed for frameworks $(ProjectFramework).
[15:30:27 WRN] Analysis of project Library\Library.csproj failed for frameworks $(ProjectFramework).
[15:30:27 WRN] Analysis of project Application\Application.csproj failed for frameworks $(ProjectFramework).
[15:30:27 INF] Analysis complete.
[15:30:27 INF] Time Elapsed 00:00:05.5424365
Stryker.NET failed to mutate your project. For more information see the logs below:


No project references found. Please add a project reference to your test project and retry.
```

When changing the `<TargetFramework>` in each .csproj file to "net8.0", it works:

```bash
> dotnet stryker -tp .\Tests\Tests.csproj -p Library\Library.csproj
   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 4.6.0

[15:35:36 INF] Analysis starting.
[15:35:36 INF] Identifying projects to mutate in Repro.sln. This can take a while.
[15:35:39 WRN] Analysis of project Application\Application.csproj failed for frameworks $(ProjectFramework).
[15:35:40 INF] Found project Library\Library.csproj to mutate.
[15:35:40 INF] Analysis complete.
[15:35:40 INF] Building solution Repro.sln
[15:35:40 INF] Building project Repro.sln using dotnet build Repro.sln (directory D:\Projects\stryker-repro.)
[15:35:57 INF] Number of tests found: 4 for project \Library\Library.csproj. Initial test run started.
[15:36:00 INF] 4 mutants created
[15:36:00 INF] Capture mutant coverage using 'CoverageBasedTest' mode.
Hint: by passing "--open-report or -o" the report will open automatically and update the report in real-time.
[15:36:01 INF] 2     mutants got status Ignored.      Reason: Removed by block already covered filter
[15:36:01 INF] 2     total mutants are skipped for the above mentioned reasons
[15:36:01 INF] 2     total mutants will be tested
██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████100.00% │ Testing mutant 2 / 2 │ K 2 │ S 0 │ T 0 │ ~0m 00s │                                                                                                          00:00:01
Killed:   2
Survived: 0
Timeout:  0
```

**Comment by dupdob** [comment]
thanks for the repro project. I just downloaded it and will try it. I will revert as soon as I figured out the problem

**Comment by dupdob** [comment]
I have good and bad news: I fail to reproduce the problem with this solution. Everything works perfectly 'out of the box'. 

Meaning it looks like it is related to your setup, maybe the specific SDK that is used... 

Could you try running your repro project using the `--dev-mode` option? this will dump the project analysis logs and it could give us some hint.

Note: I am running the latest in dev version, but there have been no change regarding project analysis since 4.6.0, so I am pretty confident result would be the same with the official 4.6.0

**Comment by willseward** [comment]
Ok. I ran with `--dev-mode` with the non-working configuration, and this was the result:

```text
[08:00:14 WRN] Project analysis failed. The MsBuild log is below.
[08:00:14 INF] Started process 33728: "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin\MSBuild.exe"  /noconsolelogger /restore /target:Clean;Build /property:DesignTimeBuild="true";BuildingProject="false";BuildProjectReferences="false";SkipCompilerExecution="true";DisableRarCache="true";AutoGenerateBindingRedirects="false";CopyBuildOutputToOutputDirectory="false";CopyOutputSymbolsToOutputDirectory="false";CopyDocumentationFileToOutputDirectory="false";ComputeNETCoreBuildOutputFiles="false";SkipCopyBuildProduct="true";AddModules="false";UseCommonOutputDirectory="true";GeneratePackageOnBuild="false";NonExistentFile="__NonExistentSubDir__\__NonExistentFile__";ResolveNuGetPackages="true";SolutionDir="XXXXX\\" /l:BuildalyzerLogger,"D:\packages\nuget\dotnet-stryker\4.6.0\tools\net8.0\any\Buildalyzer.Logger.dll";9908;False /noAutoResponse "XXXXX\Library\Library.csproj"
Microsoft (R) Build Engine version 16.11.6+a918ceb31 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.
Version 8.0.411 of the .NET SDK requires at least version 17.8.3 of MSBuild. The current available version of MSBuild is 16.11.6.22506. Change the .NET SDK specified in global.json to an older version that requires the MSBuild version currently available.The SDK 'Microsoft.NET.Sdk' specified could not be found.Process 33728 exited with code 1
Started process 16996: "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin\MSBuild.exe"  /noconsolelogger /restore /target:Clean;Build /property:DesignTimeBuild="true";BuildingProject="false";BuildProjectReferences="false";SkipCompilerExecution="true";DisableRarCache="true";AutoGenerateBindingRedirects="false";CopyBuildOutputToOutputDirectory="false";CopyOutputSymbolsToOutputDirectory="false";CopyDocumentationFileToOutputDirectory="false";ComputeNETCoreBuildOutputFiles="false";SkipCopyBuildProduct="true";AddModules="false";UseCommonOutputDirectory="true";GeneratePackageOnBuild="false";NonExistentFile="__NonExistentSubDir__\__NonExistentFile__";ResolveNuGetPackages="true";SolutionDir="XXXXX\\" /l:BuildalyzerLogger,"D:\packages\nuget\dotnet-stryker\4.6.0\tools\net8.0\any\Buildalyzer.Logger.dll";7216;False /noAutoResponse "XXXXX\Library\Library.csproj"
Microsoft (R) Build Engine version 16.11.6+a918ceb31 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.
Version 8.0.411 of the .NET SDK requires at least version 17.8.3 of MSBuild. The current available version of MSBuild is 16.11.6.22506. Change the .NET SDK specified in global.json to an older version that requires the MSBuild version currently available.The SDK 'Microsoft.NET.Sdk' specified could not be found.Process 16996 exited with code 1
```

Swapping the `$(ProjectFramework)` for the string `net8.0` does not emit that error, as we expected. MSBuild 16 is curious choice given that I have VS 2022 also installed and that it doesn't support net8.0. Fwiw, I have MSBuild 16 installed through the VS 2019 Build Tools package, not through VS directly.

**Comment by dupdob** [comment]
thanks, so this is a problem of MsBuild version selection. Now I wonder why it does work when the target framework is specified. This looks so unrelated....

**Comment by willseward** [comment]
If the MSBuild version selection isn't part of Stryker, then I completely agree. Seems like this is a "me" problem :D.

At least this has been documented, and could assist someone else having this issue in the future.

**Comment by dupdob** [comment]
I dug further into this and I think I fully understand the problem now:

### issue summary
Stryker may fail to analyze or build a project when the target framework is not explicitly specified in the relevant project file.

### underlying cause: Buildalyzer automatic build environment selection logic:
Buildalyzer may fail to automatically select the proper build environment if the target framework is specified in the project file.

### fix: won't be fixed
This is a more of an enhancement than a fix: Buildalyzer should replicate the search and parsing logic of 'Directory.Build.Props' as done by latest SDKs. This requires significant work and would not be a future proof solution so I see no path forward on this.
This is to be understood as a limitation of automatic build environment resolution.

### therorical solution: via configuration
The proper way to address this is to ask the user the specify which msbuild version to use. Stryker offers the `msbuild-path` option to allow one to select the version of it's choosing. Alas, this option is not forwarded to Buildalyzer, if I understand the code correctly. I will open an issue for this.

### workaround: use `targetframework` option
That being said, forcing user to specify the msbuild path is not practical as the exact path will depend on the machine configuration, so Stryker should try to overcome this without machine specific configuration.

One way to deal with that is to restore the behavior from earlier version when the target framework was forwarded to Buildalyzer. As such, it picks the proper msbuild version and everything runs smoothly afterward.

I have a PR for that which is almost ready: the targetframework option is forwarded to Buildalyzer when automatic mode fails.

for information cc @rouke-broersma @richardwerkman @Mobrockers
